integration-tests/src/test/java/com/datastax/oss/driver/mapper/QueryProviderIT.java

```diff
@@ -0,0 +1,231 @@
+
+  private static SensorDao dao;
+
+  @BeforeClass
+  public static void setup() {
+    CqlSession session = sessionRule.session();
+
+    session.execute(
+        SimpleStatement.builder(
+                "CREATE TABLE sensor_reading(sensor_id int, month int, day int, value double, "
+                    + "PRIMARY KEY (sensor_id, month, day)) "
+                    + "WITH CLUSTERING ORDER BY (month DESC, day DESC)")
+            .setExecutionProfile(sessionRule.slowProfile())
+            .build());
+
+    SensorMapper mapper = new QueryProviderIT_SensorMapperBuilder(session).build();
+    dao = mapper.sensorDao(sessionRule.keyspace());
+  }
+
+  @Test
+  public void should_invoke_query_provider() {
+    SensorReading readingFeb3 = new SensorReading(1, 2, 3, 9.3);
+    SensorReading readingFeb2 = new SensorReading(1, 2, 2, 8.6);
+    SensorReading readingFeb1 = new SensorReading(1, 2, 1, 8.7);
+    SensorReading readingJan31 = new SensorReading(1, 1, 31, 8.2);
+    dao.save(readingFeb3);
+    dao.save(readingFeb2);
+    dao.save(readingFeb1);
+    dao.save(readingJan31);
+
+    assertThat(dao.findSlice(1, null, null).all())
+        .containsExactly(readingFeb3, readingFeb2, readingFeb1, readingJan31);
+    assertThat(dao.findSlice(1, 2, null).all())
+        .containsExactly(readingFeb3, readingFeb2, readingFeb1);
+    assertThat(dao.findSlice(1, 2, 3).all()).containsExactly(readingFeb3);
+  }
+
+  @Mapper
+  public interface SensorMapper {
+    @DaoFactory
+    SensorDao sensorDao(@DaoKeyspace CqlIdentifier keyspace);
+  }
+
+  @Dao
+  public interface SensorDao {
+    @QueryProvider(providerClass = FindSliceProvider.class, entityHelpers = SensorReading.class)
+    PagingIterable<SensorReading> findSlice(int id, Integer month, Integer day);
+
+    @Insert
+    void save(SensorReading reading);
+  }
+
+  public static class FindSliceProvider {
+    private final Session session;
+    private final EntityHelper<SensorReading> sensorReadingHelper;
+    private final Select selectStart;
+
+    public FindSliceProvider(
+        MapperContext context, EntityHelper<SensorReading> sensorReadingHelper) {
+      this.session = context.getSession();
+      this.sensorReadingHelper = sensorReadingHelper;
+      this.selectStart =
+          sensorReadingHelper.selectStart().whereColumn("sensor_id").isEqualTo(bindMarker());
+    }
+
+    public PagingIterable<SensorReading> findSlice(int id, Integer month, Integer day) {
+      if (month == null && day != null) {
+        throw new IllegalArgumentException("Can't specify day if month is null");
+      }
+
+      Select select = this.selectStart;
+      if (month != null) {
+        select = select.whereColumn("month").isEqualTo(bindMarker());
+        if (day != null) {
+          select = select.whereColumn("day").isEqualTo(bindMarker());
+        }
+      }
+      PreparedStatement preparedStatement =
+          session.execute(new DefaultPrepareRequest(select.build()), PrepareRequest.SYNC);
```

Comment:
So far the mapper wraps the top-level `Session` interface, which will force `QueryProvider` implementations to use the generic execute method.

I'm a bit on the fence about that: on the one hand it's nice to be as generic as possible, on the other that generic signature is a pain to use, and the probability that someone would initialize the mapper with something that is not a `CqlSession` is really thin.

Reply:
All things considered I'm going to do the change to take a `CqlSession`.

Heads up that I'll force-push it on this PR.